src/main/java/net/logstash/logback/stacktrace/StackHasher.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.stacktrace;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * Utility class that generates a hash for any Java {@link Throwable error} stack
+ *
+ * @author Pierre Smeyers
+ */
+public class StackHasher {
+
+    private final StackElementFilter filter;
+
+    /**
+     * Constructs a {@link StackHasher} with the given filter
+     *
+     * @param filter filter
+     */
+    public StackHasher(StackElementFilter filter) {
+        this.filter = filter;
+    }
+
+    /**
+     * Constructs a {@link StackHasher} using {@link StackElementFilter#any()} filter
+     */
+    public StackHasher() {
+        this(StackElementFilter.any());
+    }
+
+    /**
+     * Generates a Hexadecimal hash for the given error stack
+     * <p>
+     * Two errors with the same stack hash are most probably same errors
+     */
+    public String hexHash(Throwable error) {
+        return hexHashes(error).peek();
+    }
+
+    /**
+     * Generates and returns Hexadecimal hashes for the error stack and each ancestor {@link Throwable#getCause() cause}
+     * <p>
+     * The first queue element is the stack hash for the topmost error, the next one (if any) is it's direct
+     * {@link Throwable#getCause() cause} hash, and so on...
+     */
+    public Deque<String> hexHashes(Throwable error) {
```

Comment:
As I alluded to in other comments, I don't think StackHasher should have a method that returns a `Deque`.  The Deque complicates the code a bit (since callers need to know how to correlate entries with causes), and will result in extra garbage in a code "hot path".

Instead, I would make it provide a method like this:

```java
/**
 * Generates and returns Hexadecimal hash for the given Throwable.
 *
 * @param throwable The throwable for which to calculate the hash
 * @param includeCauses when true, each cause will be considered when calculating the hash.
 *                      when false, only the top level throwable will be considered in the hash.
 */
public String computeHash(Throwable throwable, boolean includeCauses)
```


StackHashJsonProvider would call this method once, using includeCauses=true, since it only needs one value for the complete throwable chain.

ShortenedThrowableConverter would call this method using includeCauses=false for each throwable in the cause chain as it recursively traverses each throwable.